### PR TITLE
Convert SensorFusion World output to TbotsProto::Vision and ProtoLogger log it

### LINKED ISSUE #1789: Log filtered world data as protobuf
### Description of the task

There is already a function `createVision` that converts `World` to `TbotsProto::Vision`, use this method and log the filtered world to Protobuf

### Acceptance criteria

- [ ] Log filtered world data to protobuf

### PR DESCRIPTION
**Pull request by jordanzliu**
### Description

- Refactor Observer into Observer interface and ObserverBuffer as a derived class with the buffer implementation currently used
- Create SubjectObserverAdapter as both a Subject and Observer that converts the observed type into the subject type
- Use the above to log filtered world state

### Testing Done

image

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by MathewMacDougall — COMMENTED** [review]
Looks reasonable overall. Only real issue is with naming of `ObserverBuffer`. It makes me think it's an actual buffer of some type.

Is the only reason you made the `Observer` an interface and moved the previous functionality to `ObserverBuffer` to avoid extra/unneeded functionality in the `ObserverSubjectAdapter`? Might just be a bit simpler to leave the Observer class as-is and have `ObserverSubjectAdapter` inherit from it directly. It would end up with a few more functions but nothing too orthogonal.

**Comment by jordanzliu** [comment]
@MathewMacDougall From a pure OO design standpoint, it makes sense for `Observer` to be a pure virtual class, but I suppose having orthogonal functionality in `Subject` would be "less bad" than having multiple levels of inheritance. Will make the change

**Comment by MathewMacDougall** [comment]
Yea that's fair. I think it was mostly the resultant naming that I didn't like vs. the technical aspect. Curious what others think is more readable / maintainable as well.

File: src/software/multithreading/observer_subject_adapter.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
+#pragma once
+#include <functional>
+
+#include "software/multithreading/observer.h"
+#include "software/multithreading/subject.h"
+
+/**
+ * ObserverSubjectAdapter is a class made to "adapt" a type emitted by a Subject to
```
Comment (jonathanlew) [inline src/software/multithreading/observer_subject_adapter.h:8]: could add this design pattern link: https://refactoring.guru/design-patterns/adapter

**Review by jonathanlew — COMMENTED** [review]
Pretty clean!

**Review by akhilveeraghanta — COMMENTED** [review]
lgtm!

**Review by jonathanlew — APPROVED** [review]

**Review by akhilveeraghanta — APPROVED** [review]

**Pull request merged**